games/whosaidit/whosaidit.py

```diff
@@ -0,0 +1,106 @@
+    return winners
+
+
+async def play(bot, ctx, players):
+    """
+    play is the main gameplay loop.
+
+    It randomly selects from config["channels"] a config["history_limit"]
+    number of messages, somewhere between the channel's creation time and
+    now.
+    
+    All messages that have an author of config["required_role"] and a message
+    length greater than config["min_num_words"] are added for contention.
+
+    Then, for all members in config["author_channel"], randomly select 4,
+    and add the message's author to choose from.
+    """
+    messages = list()
+    authors = set()
+
+    possible_channels = [(int(channel), weight) for channel, weight in config["channels"].items()]
+    # random.choices() returns a k-sized list. We only want one, so we select the first element.
+    # Since these channels are stored as tuples for (channel, weight), we select the first element.
+    target_channel = bot.get_channel(random.choices(possible_channels, weights=[c[1] for c in possible_channels], k=1)[0][0])
+    target_created_at = target_channel.created_at
+    current_time = datetime.datetime.now(tz=None)
+
+    # If there's ever a case that history() will return 0 messages,
+    # we need to randomly choose another time and do the same calculation.
+    while len(messages) == 0:
+        random_time = target_created_at + \
+            datetime.timedelta(seconds=random.randint(0,
+            int((current_time - target_created_at).total_seconds())))
+        async for message in target_channel.history(limit = config["history_limit"], after=random_time, oldest_first=True):
+            try:
+                if bot.get_guild(config["guild_id"]).get_role(config["required_role"]) in message.author.roles and len(message.content.split(" ")) >= config["min_num_words"]:
+                    messages.append(message)
+            except Exception as e:
+                print(f"Failed polling message: {e}")
+                continue
+
+    message_to_guess = random.choice(messages)
+    channel_members = set(bot.get_channel(config["author_channel"]).members)
+    channel_members = set([member for member in channel_members if not member.bot])
+    channel_members.remove(message_to_guess.author)
+    authors.update(random.sample(channel_members, 4))
+    authors.add(message_to_guess.author)
+    random.shuffle(list(authors))
+
+    choices = {choice + 1: author for choice, author in enumerate(authors)}
+    answer = list(choices.keys())[list(choices.values()).index(message_to_guess.author)]
+
+    await display(ctx, message_to_guess, choices)
+    guesses = await get_choice(bot, ctx, choices, players)
+    winners = []
+    for player, guess in guesses.items():
+        if guess == answer:
+            winners.append(bot.get_user(player))
+            await ctx.send(f"This quote was from {message_to_guess.author.name} on <t:{int(message_to_guess.created_at.timestamp())}:d>!")
+            return winners
```

Comment:
This results in a bug.

If we do not have _any_ correct guesses, `winners` is never returned, and we end in a bad state. A potential way is to return empty winners, but we can probably refactor for better logic here.

```suggestion
    for player, guess in guesses.items():
        if guess == answer:
            winners.append(bot.get_user(player))
            await ctx.send(f"This quote was from {message_to_guess.author.name} on <t:{int(message_to_guess.created_at.timestamp())}:d>!")
            return winners
    return winners
```

Reply:
Reworked to return empty winners, that's how other non zero-sum chowderbot games work for coin/cgr calculations